### PR TITLE
Bidirectional RNNs

### PR DESCRIPTION
**Pull request by iskandr**
This is an implementation of bidirectional RNNs using the `go_backwards` suggestion from https://github.com/fchollet/keras/issues/418. 

* Added a `go_backwards` parameter to every RNN class, which is then passed to `theano.scan`

* Added a layer called `recurrent.Bidirectional`, which is parameterized by an RNN class as well arguments to pass to that RNN class when creating forward and backward instances of it. One choice I'm not sure was the best: the `output_dim` of a `Bidirectional` layer is the sum of the output dims of the forward and backward models (and thus each model gets a hidden state of size `output_dim // 2`). 

* Added a unit test which trains a bidrectional LSTM on a trivial task (detecting whether a character is present or absent in a small set of sequences). 

* Changed the IMDB LSTM example to use a bidirectional LSTM

This is my first attempt at a contribution to Keras and I'm not sure if I implemented enough (or too many) of the methods needed to make a fully functioning Layer, any and all tips & feedback very welcome.

### CODE REVIEW

**Comment by jedi00** [comment]
It works when the bidirectional layer is not the first layer, but breaks if it is. 
It prompts "theano.compile.function_module.UnusedInputError: theano.function was asked to create a function computing outputs given certain inputs, but the provided input variable at index 0 is not part of the computational graph needed to compute the outputs: <TensorType(float64, 3D)>.
"
No clue where it went wrong for now.

**Comment by iskandr** [comment]
I copied this bit blindly without really understanding how the first layer in the network plugs into the Theano compute graph:

```
       # create input tensor in case this is the first layer in a network
        self.input = T.tensor3()
```

I'm not sure what I should be doing here instead, any ideas @jedi00 or @fchollet?

**Comment by jedi00** [comment]
@iskandr : There's no problem with this line. Keras will use the first layer's .input atrribute to figure out the input shape of the whole model. 
BTW, did you try my example class inheritance code #418 ? It seems the same problem. I've checked the computation graph of each para of the model, they're all the same with the ones when LSTM is used instead. This is really gnawing.

**Comment by jedi00** [comment]
Finally I captured the bug! The problem lies in the data input path of the the computational graph. When the bidirectional layer is the first layer of the whole model, the input of the underlying forward rnn & backward rnn are not connected properly.
I've submitted a new pull request #465 , resolved this problem.

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -707,19 +759,119 @@ def get_output(self, train=False):
...
+        # In either case, concatenate the two outputs to get a final dimension
+        # of output_dim * 2
+        return T.concatenate(
+            [forward_output, backward_output],
+            axis=-1)
+
+    def get_output_mask(self, train=False):
+        if self.return_sequences:
+            mask = self.get_input_mask(train)
+            print(mask, mask.ndim)
```
Comment (wxs) [inline keras/layers/recurrent.py]: This print statement is probably left over from debugging, no?

**Comment by wxs** [comment]
Any idea how this behaves with masking? I don't fully understand the "go_backwards" flag of Theano, but its relationship with the mask seems dangerous.

My concern is, in every recurrent layer we generate a mask with this function: `padded_mask = self.get_padded_shuffled_mask(train, X, pad=1)`. This prepends a pad to the beginning of the mask so that at each timestep of `scan` you can view the mask of the previous timestep.

It seems like with `go_backwards=True` you would want to instead *append* the pad (i.e. to the end of the array).

File: keras/layers/recurrent.py (hunk 1 above)
Comment (jedi00) [inline keras/layers/recurrent.py]: @wxs : this get_output_mask() is from @iskandr .

**Comment by jedi00** [comment]
@wxs: According to theano documentation, when go_backwards = True, scan() will iterate with all input sequences in backwards direction. After discussing the padding mechansim with you, I just made modification to make the padding comply with the 'go_backwards' parameter @ #465 .

Sadly I still can't make the Bidirectional class work with other RNN class.  I built a model like:
```
model = Sequential()
model.add(Bidirectional(JZS1, 28, 28))
model.add(Dropout(0.1))
model.add(Dense(28, nb_classes, activation='softmax'))
```
however, error raised as
```
Traceback (most recent call last):
  File "C:\Python34\lib\site-packages\theano-0.7.0-py3.4.egg\theano\compile\function_module.py", line 594, in __call__
    outputs = self.fn()
ValueError: Input dimension mis-match. (input[0].shape[1] = 28, input[1].shape[1] = 14)
```
Any suggestion?

**Comment by wxs** [comment]
@jedi00 don't you need a `TimeDistributedDense` there?

**Comment by jedi00** [comment]
@wxs: `Bidirectional` returns non-sequence, so a `Dense` layer is used. If there is no 'Dropout' layer, the model works fine; the error is raised only when another RNN layer is appended after `Bidirectional' layer.

**Comment by iskandr** [comment]
@jedi00 @wxs @fchollet I haven't followed the past month of activity closely, do any of the recent PRs contain examples of layers which take other layers as arguments? I'm not sure what goes wrong with dropout following a `Bidirectional` but it might help to see an example of something that actually works.

**Pull request closed without merging**